src/compile/data/aggregate.ts

```diff
@@ -0,0 +1,134 @@
+import {AggregateOp} from '../../aggregate';
+import {Channel} from '../../channel';
+import {SOURCE} from '../../data';
+import {field, FieldDef} from '../../fielddef';
+import {keys, vals, reduce, hash, Dict, StringSet, allSame, empty} from '../../util';
+import {VgData} from '../../vega.schema';
+
+import {FacetModel} from './../facet';
+import {LayerModel} from './../layer';
+import {Model} from './../model';
+
+import {DataComponent, AggregateComponent} from './data';
+
+
+export namespace aggregate {
+  function addDimension(dims: { [field: string]: boolean }, fieldDef: FieldDef) {
+    if (fieldDef.bin) {
+      dims[field(fieldDef, { binSuffix: '_start' })] = true;
+      dims[field(fieldDef, { binSuffix: '_mid' })] = true;
+      dims[field(fieldDef, { binSuffix: '_end' })] = true;
+      dims[field(fieldDef, { binSuffix: '_range' })] = true;
+    } else {
+      dims[field(fieldDef)] = true;
+    }
+    return dims;
+  }
+
+  export function parseUnit(model: Model): AggregateComponent {
+    /* string set for dimensions */
+    let dims: StringSet = {};
+
+    /* dictionary mapping field name => dict set of aggregation functions */
+    let meas: Dict<StringSet> = {};
+
+    model.forEach(function (fieldDef: FieldDef, channel: Channel) {
+      if (fieldDef.aggregate) {
+        if (fieldDef.aggregate === AggregateOp.COUNT) {
+          meas['*'] = meas['*'] || {};
+          /* tslint:disable:no-string-literal */
+          meas['*']['count'] = true;
+          /* tslint:enable:no-string-literal */
+        } else {
+          meas[fieldDef.field] = meas[fieldDef.field] || {};
+          meas[fieldDef.field][fieldDef.aggregate] = true;
+        }
+      } else {
+        addDimension(dims, fieldDef);
+      }
+    });
+
+    return {
+      dimensions: dims,
+      measures: meas
+    };
+  }
+
+  export function mergeMeasures(parentMeasures: Dict<Dict<boolean>>, childMeasures: Dict<Dict<boolean>>) {
+    for (const field in childMeasures) {
+      if (childMeasures.hasOwnProperty(field)) {
+        // when we merge a measure, we either have to add an aggregation operator or even a new field
+        const ops = childMeasures[field];
+        for (const op in ops) {
+          if (ops.hasOwnProperty(op)) {
+            if (!(field in parentMeasures)) {
+              parentMeasures[field] = {};
+            }
+            parentMeasures[field][op] = true;
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Add facet fields as dimensions.
+   */
+  export function parseFacet(model: FacetModel, aggregateComponent: AggregateComponent) {
+    aggregateComponent.dimensions = model.reduce(addDimension, aggregateComponent.dimensions);
+  }
+
+  /**
+   * Merges the aggregates from the children into the data component.
+   */
+  export function merge(dataComponent: DataComponent, childDataComponents: DataComponent[]) {
```

Comment:
Why is this function correct with facet? 

For facet, the dimension of the parent do not have to equal the child.  But all child's dimension (and source) have to be equal.  

If all children have the same dimensions, we can merge both dimensions and measures.  The new dimension = child's dimension union with the parent's dimension.  Basically, `facet` provides additional dimensions to the children.   

On the other hand, if the dimensions for children are not the same, we have to append the facet's dimensions into dimensions of each child that does not have its own data source so that data are correctly splitted across facets. 


For non-facet parents, I assume there should always be no dimensions for the parents?